test/e2e/storage/csi_volumes.go

```diff
@@ -367,57 +369,96 @@ func initCSIgcePD(f *framework.Framework, config framework.VolumeTestConfig) csi
-	csiControllerRoleBinding(cs, config, false, role, g.controllerServiceAccount)
-	deployGCEPDCSIDriver(cs, config, false /* teardown */, f, g.nodeServiceAccount, g.controllerServiceAccount)
+	cleanup, err := g.f.CreateFromManifest(func(object interface{}) {
+		patchCSIDeployment(g.f, "com.google.csi.gcepd", "gce-driver", "csi-external-provisioner", g.config.ServerNodeName, object)
+	},
+		"test/e2e/testing-manifests/storage-csi/driver-registrar/rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/external-attacher/rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/external-provisioner/rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/csi-controller-rbac.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/node_ds.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml",
+		"test/e2e/testing-manifests/storage-csi/gce-pd/controller_service.yaml",
+	)
+	g.cleanup = cleanup
+	if err != nil {
+		framework.Failf("deploying csi hostpath driver: %v", err)
+	}
 }
 
 func (g *gcePDCSIDriver) cleanupCSIDriver() {
 	By("uninstalling gce-pd driver")
-	f := g.f
-	cs := f.ClientSet
-	config := g.config
-	deployGCEPDCSIDriver(cs, config, true /* teardown */, f, g.nodeServiceAccount, g.controllerServiceAccount)
-	csiClusterRoleBindings(cs, config, true /* teardown */, g.controllerServiceAccount, g.controllerClusterRoles)
-	csiClusterRoleBindings(cs, config, true /* teardown */, g.nodeServiceAccount, g.nodeClusterRoles)
-	utils.PrivilegedTestPSPClusterRoleBinding(cs, config.Namespace, true, /* teardown */
-		[]string{g.controllerServiceAccount.Name, g.nodeServiceAccount.Name})
-	role := csiControllerRole(cs, config, true)
-	csiControllerRoleBinding(cs, config, true, role, g.controllerServiceAccount)
-	csiServiceAccount(cs, config, "gce-controller", true /* teardown */)
-	csiServiceAccount(cs, config, "gce-node", true /* teardown */)
+	if g.cleanup != nil {
+		g.cleanup()
+	}
+}
+
+// TODO (?): the storage.csi.image.version and storage.csi.image.registry
+// settings are ignored. We could patch the image definitions.
+// TODO: way too many parameters...
+func patchCSIDeployment(f *framework.Framework, provisioner, driverContainerName, provisionerContainerName, serverNodeName string, object interface{}) {
+	patchVolumes := func(volumes []v1.Volume) {
+		for i := range volumes {
+			volume := &volumes[i]
+			if volume.HostPath != nil {
+				// Update paths like /var/lib/kubelet/plugins/<provisioner>.
+				path := &volume.HostPath.Path
+				if strings.HasSuffix(*path, "/"+provisioner) {
+					*path = *path + "-" + f.UniqueName
+				}
+			}
+		}
+	}
+
+	patchContainers := func(containers []v1.Container) {
+		for i := range containers {
+			container := &containers[i]
+			for e := range container.Args {
+				// Inject test-specific provider name into paths (--kubelet-registration-path).
+				container.Args[e] = strings.Replace(container.Args[e], "/"+provisioner+"/", "/"+provisioner+"-"+f.UniqueName+"/", 1)
```

Comment:
Is this a duplicate of L445?

Reply:
The comment said it's updating kubelet registration path, but the code looks like it's updating the provisioner argument

Reply:
The code does what the comment says:
`--kubelet-registration-path=/var/lib/kubelet/plugins/csi-hostpath/csi.sock` becomes `--kubelet-registration-path=/var/lib/kubelet/plugins/csi-hostpath-e2e-tests-csi-volumes-q7vzj/csi.sock`.

I'll give that as full example in the comment.

The comment from above for `--drivername` also applies here: this can work for multiple different drivers because they are likely to follow this convention (or even have to, in this case?).

Reply:
Isn't "provisioner" and "driver name" synonymous? Driver registrar expects a `--provisioner` parameter, while hostpath takes a `--drivername`, and both are the same string as far as I understand it.

Please have a look at the revised code (https://github.com/kubernetes/kubernetes/pull/69868/files#diff-10be18decd977f7f8fe7c0c02e2cbf05), perhaps it is clearer there. If not, please let me know how to improve it.

Reply:
There should not be a separate provisioner name.  IMO it was a mistake to make provisioner name configurable.  It should really be equal to the driver name.  But I think overall, we should be referring to driver name, not provisioner name.  Not all drivers may have a provisioner.

Reply:
(I'm not seeing a provisioner parameter do the driver-registrar)

Reply:
It takes `--kubelet-registration-path` which happens to get a path which contains "the string that is sometimes the provider name and sometimes the driver name". If it's better to call that string "driver name" then I'll change the code accordingly.

Reply:
Done, and added comments explaining that the provisioner name is expected to be the same.